docs: migrate 'how to set up classic confinement' (#5345)
canonical/snapcraft: docs/how-to/craft/enable-classic-confinement.rst

        :caption: snapcraft.yaml
        :language: yaml
        :lines: 2-

Make plugin
~~~~~~~~~~~

`makefile-lib-example
<https://github.com/snapcraft-doc-samples-unofficial/makefile-lib-example>`_ is a
project with a main part that uses the Make plugin.

.. collapse:: makefile-lib-example project file

    .. literalinclude:: code/example-classic-confinement-make.yaml
        :caption: snapcraft.yaml
        :language: yaml
        :lines: 2-

CMake plugin
~~~~~~~~~~~~

`cmake-classic-example <https://github.com/snapcraft-docs/cmake-classic-example>`_ is a
project with a main part that uses the CMake plugin.

.. collapse:: cmake-classic-example project file

    .. literalinclude:: code/example-classic-confinement-cmake.yaml
        :caption: snapcraft.yaml
        :language: yaml
        :lines: 2-

Python plugin
~~~~~~~~~~~~~

`python-ctypes-example <https://github.com/snapcraft-docs/python-ctypes-example>`_ is a
project with a main part that uses the Python plugin.

.. collapse:: python-ctypes-example project file

    .. literalinclude:: code/example-classic-confinement-python.yaml
        :caption: snapcraft.yaml
        :language: yaml
        :lines: 2-

> Patch Python ctypes to load system libraries

Author: I'm not confident in any of the language of this section.

^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

If your app uses ctypes to access system libraries, it must be bundled with patched
ctype files:

> 1. Stage the system libraries in the main part. For the core22 base, the packages are:

Reviewer: You could rephrase this to provide a some background on why Python 3.10 is used for core22:
```suggestion
1. Stage the system libraries in the main part. The core22 base uses python 3.10, so the packages are:
```

Author: [d49dfb3](https://github.com/canonical/snapcraft/pull/5345/commits/d49dfb3e5f8e44dca56906c51ed9a8f43e9f83ac)

   .. code-block:: yaml
       :caption: snapcraft.yaml

       stage-packages:
         - libpython3.10-minimal
         - libpython3.10-stdlib

2. Override the part's build to also run a shell script that patches the ctype files:

   .. code-block:: yaml
       :caption: snapcraft.yaml

       override-build: |

>          snapcraftctl build
>          $SNAPCRAFT_PROJECT_DIR/snap/local/patch-ctypes.sh

Reviewer: Both work but these are the modern calls:
```suggestion
         craftctl build
         $CRAFT_PROJECT_DIR/snap/local/patch-ctypes.sh
```

Author: [d49dfb3](https://github.com/canonical/snapcraft/pull/5345/commits/d49dfb3e5f8e44dca56906c51ed9a8f43e9f83ac)

   The plugin now builds the part like normal before running the patch script.

3. Build the snap. During build, the script looks for core Python modules that need to
   be patched and refer to libraries in the base snap.

.. collapse:: Patch script for ctype files

>     .. literalinclude:: code/example-classic-confinement-python-ctype-patches.sh

Reviewer: This patching script goes hand-in-hand with the patch itself: https://github.com/snapcraft-docs/python-ctypes-example/blob/main/snap/local/patches/ctypes_init.diff

Should we include them both here?

Author: I didn't notice it was grabbing a diff from local files. I think it's better instead to keep those bits hosted, otherwise we may as well include the whole project inside the docs. [d49dfb3](https://github.com/canonical/snapcraft/pull/5345/commits/d49dfb3e5f8e44dca56906c51ed9a8f43e9f83ac)

Enable classic confinement
==========================

When a snap needs system resources beyond what strict confinement and the available
`interfaces <https://snapcraft.io/docs/supported-interfaces>`_ can provide, it
needs `classic confinement <https://snapcraft.io/docs/classic-confinement>`_.

This page provides guidance on how to enable classic confinement, as well as example
solutions for snaps that have extra plugin-specific configurations to accommodate the
confinement.

Set the confinement level
-------------------------

You can enable classic confinment with a single key in the project file:

.. code-block:: yaml
    :caption: snapcraft.yaml

    confinement: classic

> When built, the snap now has admin access to the system.

Author: > admin access

Not sure how to better phrase this. The original doc said "full system access", which to me implies side effects that aren't true. I welcome suggestions

Reviewer: I think "admin access" is too close to "root access" and the snap doesn't have root access.

It could be phrased like "the snap is no longer sandboxed" or "snapd runs the snap without any sandboxing or permission restrictions".

I like the sentence in the original doc that compares classic confinement to running "a traditionally packaged application", although it could be phrased like "a local executable file".

Author: I went with "unsandboxed access to the system": [d49dfb3](https://github.com/canonical/snapcraft/pull/5345/commits/d49dfb3e5f8e44dca56906c51ed9a8f43e9f83ac)

Reviewer: I like it.

~~~~~~~~~~~

`makefile-lib-example
<https://github.com/snapcraft-doc-samples-unofficial/makefile-lib-example>`_ is a
project with a main part that uses the Make plugin.

.. collapse:: makefile-lib-example project file

    .. literalinclude:: code/example-classic-confinement-make.yaml
        :caption: snapcraft.yaml
        :language: yaml
        :lines: 2-

CMake plugin
~~~~~~~~~~~~

`cmake-classic-example <https://github.com/snapcraft-docs/cmake-classic-example>`_ is a
project with a main part that uses the CMake plugin.

.. collapse:: cmake-classic-example project file

    .. literalinclude:: code/example-classic-confinement-cmake.yaml
        :caption: snapcraft.yaml
        :language: yaml
        :lines: 2-

Python plugin
~~~~~~~~~~~~~

`python-ctypes-example <https://github.com/snapcraft-docs/python-ctypes-example>`_ is a
project with a main part that uses the Python plugin.

.. collapse:: python-ctypes-example project file

    .. literalinclude:: code/example-classic-confinement-python.yaml
        :caption: snapcraft.yaml
        :language: yaml
        :lines: 2-

Patch Python ctypes to load system libraries
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> If your app uses ctypes to access system libraries, it must be bundled with a patched
> version of the ctype files:

Author: ```suggestion
If your app uses ctypes to access system libraries, it must be bundled with patched
ctype files:
```

1. Stage the system libraries in the main part. For the core22 base, the packages are:

   .. code-block:: yaml
       :caption: snapcraft.yaml

       stage-packages:
         - libpython3.10-minimal
         - libpython3.10-stdlib

2. Override the part's build to also run a shell script that patches the ctype files:

   .. code-block:: yaml
       :caption: snapcraft.yaml

       override-build: |
         snapcraftctl build
         $SNAPCRAFT_PROJECT_DIR/snap/local/patch-ctypes.sh

   The plugin now builds the part like normal before running the patch script.

3. Build the snap.

> 4. Build the snap. During build, the script looks for core Python modules that need to

Author: > core Python modules

This comes from the original doc. I'm not confident in these terms. Is there a more accurate phrase?
